src/vizier/services/query_broker/script_runner/config_map_source.go

```diff
@@ -20,111 +20,161 @@ package scriptrunner
 
 import (
 	"context"
+	"reflect"
+	"runtime"
+	"sync/atomic"
 	"time"
 
 	log "github.com/sirupsen/logrus"
 	"gopkg.in/yaml.v2"
-	v1 "k8s.io/api/core/v1"
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/watch"
-	clientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	informercorev1 "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	listercorev1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
 
 	"px.dev/pixie/src/shared/cvmsgspb"
 	"px.dev/pixie/src/utils"
 )
 
 // ConfigMapSource pulls cron scripts from config maps.
 type ConfigMapSource struct {
-	stop   func()
-	client clientv1.ConfigMapInterface
+	stop     func()
+	informer informercorev1.ConfigMapInformer
 }
 
 // NewConfigMapSource constructs a [Source] that extracts cron scripts from config maps with the label "purpose=cron-script".
 // Each config map must contain
 //   - a script.pxl with the pixel script
 //   - a configs.yaml which will be stored in the Configs field of [cvmsgspb.CronScript]
 //   - a cron.yaml that contains a "frequency_s" key
-func NewConfigMapSource(client clientv1.ConfigMapInterface) *ConfigMapSource {
-	return &ConfigMapSource{client: client}
+func NewConfigMapSource(client kubernetes.Interface, namespace string) *ConfigMapSource {
+	return &ConfigMapSource{
+		informer: informers.NewSharedInformerFactoryWithOptions(
+			client,
+			12*time.Hour,
+			informers.WithNamespace(namespace),
+			informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+				options.LabelSelector = "purpose=cron-script"
+			}),
+		).Core().V1().ConfigMaps(),
+	}
 }
 
 // Start watches for updates to matching configmaps and sends resulting updates on updatesCh.
-func (source *ConfigMapSource) Start(baseCtx context.Context, updatesCh chan<- *cvmsgspb.CronScriptUpdate) (map[string]*cvmsgspb.CronScript, error) {
-	options := metav1.ListOptions{LabelSelector: "purpose=cron-script"}
-	watcher, err := source.client.Watch(baseCtx, options)
+func (source *ConfigMapSource) Start(_ context.Context, updatesCh chan<- *cvmsgspb.CronScriptUpdate) (map[string]*cvmsgspb.CronScript, error) {
+	stopCh := make(chan struct{})
+	isInitialized := &atomic.Bool{}
+	_, err := source.informer.Informer().AddEventHandler(configMapEventHandlers(isInitialized, updatesCh))
+	if err != nil {
+		return nil, err
+	}
+	go source.informer.Informer().Run(stopCh)
+	waitUntilSynced(source.informer.Informer())
```

Comment:
Following https://blog.dsb.dev/posts/creating-dynamic-informers/, it may be a little nicer to use `cache.WaitForCacheSync(ctx.Done(), informer.HasSynced)` which seems more conventional.